coreax/solvers/recombination.py

```diff
@@ -0,0 +1,609 @@
+            # Propagate centroid coresubset weights to the underlying weights for each
+            # centroid, as defined by `centroid_indices`.
+            weight_update_indices = centroid_indices[coresubset_indices]
+            weight_update = coresubset_weights / centroid_weights[coresubset_indices]
+            updated_weights = _weights[weight_update_indices] * weight_update[..., None]
+            # Maintain a correspondence between the original data indices and the sorted
+            # indices, used to construct the balanced centroids.
+            updated_indices = _indices[weight_update_indices.reshape(-1, order="F")]
+            return updated_weights.reshape(-1, order="F"), updated_indices
+
+        in_state = (jnp.pad(weights, (0, padding)), jnp.arange(n + padding))
+        out_weights, indices = jax.lax.fori_loop(0, depth, _tree_reduce, in_state)
+        coresubset_nodes = _coresubset_nodes(
+            push_forward_nodes, out_weights, indices, self.mode
+        )
+        return Coresubset(coresubset_nodes, dataset), solver_state
+
+
+def _prepare_tree(
+    n: int, m: int, tree_reduction_factor: int = 2
+) -> tuple[int, int, int]:
+    r"""
+    Compute and apply dataset padding and compute tree count and depth.
+
+    :param n: Number of nodes
+    :param m: Number of test-functions
+    :param tree_reduction_factor: The factor by which each tree reduction step reduces
+        the number of non-zero points; the remaining number of non-zero nodes, after
+        performing recombination, is equal to `n_nodes / tree_reduction_factor`
+    :return: The required amount of padding, to allow reshaping of the nodes into equal
+        sized clusters), the tree_count (number of clusters), and the maximum tree depth
+        (number of tree_reduction iterations required to complete tree recombination)
+    """
+    tree_count = tree_reduction_factor * m
+    max_tree_depth = math.ceil(math.log(n / m, tree_reduction_factor))
+    padding = m * tree_reduction_factor**max_tree_depth - n
+    return padding, tree_count, max_tree_depth
+
+
+@jax.vmap
+def _centroid(
+    nodes: Shaped[Array, "tree_count n/tree_count m"],
+    weights: Shaped[Array, "tree_count n/tree_count"],
+) -> tuple[Shaped[Array, "n/tree_count m"], Shaped[Array, " n/tree_count"]]:
+    """
+    Compute the centroid mass and node centre (centre-of-mass).
+
+    :param nodes: A set of clustered nodes where the leading axis indexes each cluster,
+        which this function vmaps over, and the middle axis indexes each node within a
+        given cluster.
+    :param weights: A set of clustered weights associated with each node; has the same
+        index layout as the nodes.
+    :return: Cluster centroid (centre-of-mass) and total cluster mass for all clusters
+    """
+    centroid_nodes = jnp.nan_to_num(jnp.average(nodes, 0, weights))
+    centroid_weights = jnp.sum(weights)
+    return centroid_nodes, centroid_weights
```

Comment:
Move these to `util.py`, thoughts?

Reply:
This would necessitate adding unit tests for these functions

Reply:
This function is "private" here as its implementation is solely for achiving the recombination algorithms. I think making this into a proper utillity function should be left for a separate issue, considering the pull is large already.

Reply:
Happy to leave this as is

Reply:
Sorry, just to be clear, do you want a separate issue raising for this one?

Reply:
Nope fine as is, if these functions ever get reused in the future we can move them to utils then